File: solutions/ruby/palindrome-products/3/palindrome_products.rb
class Palindromes
  @@cache = {}

  def initialize(max_factor:, min_factor: 1)
    raise ArgumentError, "min must be <= max" if min_factor > max_factor
    @max_factor = max_factor
    @min_factor = min_factor
  end

  def generate
    key = [@min_factor, @max_factor]
    if @@cache[key]
      @palindromes = @@cache[key]
    else
      @palindromes = []
      palindromes_hash = {}
      (@min_factor..@max_factor).each do |i|
        (i..@max_factor).each do |j|
          product = i * j
          if palindrome?(product)
            factors_to_add = [[i, j]]
            if palindromes_hash[product]
              existing = palindromes_hash[product]
              factors_to_add.each do |factor_pair|
                existing.factors << factor_pair unless existing.factors.include?(factor_pair)
              end
            else
              @palindromes << PalindromeProduct.new(product, factors_to_add)
              palindromes_hash[product] = @palindromes.last
            end
          end
        end
      end
      @@cache[key] = @palindromes
    end
  end

  def smallest
    @palindromes.min_by(&:value) || PalindromeProduct.new(nil, [])
  end

  def largest
    @palindromes.max_by(&:value) || PalindromeProduct.new(nil, [])
  end
  
  def palindrome?(number)
    str = number.to_s
    str == str.reverse
  end
end

PalindromeProduct = Struct.new(:value, :factors)

File: solutions/csharp/palindrome-products/3/PalindromeProducts.cs
using System.ComponentModel;

public static class PalindromeProducts
{
    public static (int, IEnumerable<(int,int)>) Largest(int minFactor, int maxFactor)
    {
        var palindromeProducts = GetProducts(minFactor, maxFactor);

        var largest = palindromeProducts.Keys.Max();
        return (largest, palindromeProducts[largest]);
    }

    private static Dictionary<int, IEnumerable<(int, int)>> GetProducts(int minFactor, int maxFactor)
    {
        var palindromeProducts = new Dictionary<int, IEnumerable<(int, int)>>();
        for (var i = minFactor; i <= maxFactor; i++)
        {
            for (var j = i; j <= maxFactor; j++)
            {
                var product = (i * j).ToString();
                var digits = product.ToCharArray();
                Array.Reverse(digits);
                if (new string(digits) == product)
                {
                    var currentFactors = palindromeProducts.GetValueOrDefault(i * j, []);
                    currentFactors = currentFactors.Append((i, j));
                    palindromeProducts[i * j] = currentFactors;
                }
            }
        }

        if (palindromeProducts.Count == 0)
        {
            throw new ArgumentException();
        }

        return palindromeProducts;
    }

    public static (int, IEnumerable<(int, int)>) Smallest(int minFactor, int maxFactor)
    {
        var palindromeProducts = GetProducts(minFactor, maxFactor);
        
        var smallest = palindromeProducts.Keys.Min();
        return (smallest, palindromeProducts[smallest]);
    }
}
